File: JS/DecodeTheSlantedCiphertext.js
/**
 * @version JAVASCRIPT ECMAScript 6
 * @author MARIO SAM <eu@mariosam.com.br>
 * @see I would love to work with you instead solving web code tests: hire me!
 */

/**
 * @param {string} encodedText
 * @param {number} rows
 * @return {string}
 */

function decodeCiphertext(encodedText, rows) {
    const size = encodedText.length;
    if (rows === 1 || size <= 1) {
        return encodedText;
    }

    const cols = Math.floor(size / rows);
    let posX = 0, posY = 0;
    let ret = '';

    while (posY < cols) {
        const index = posX * cols + posY;
        ret += encodedText[index];

        posX++;
        posY++;

        if (posX >= rows) {
            posX = 0;
            posY = posY - rows + 1;
        }
    }

    return ret.replace(/\s+$/, '');
}
export { decodeCiphertext }

File: GO/DecodeTheSlantedCiphertext.go
/**
 * @version GO 1.26.0
 * @author MARIO SAM <eu@mariosam.com.br>
 * @see I would love to work with you instead solving web code tests: hire me!
 */
package GO

import "strings"

func decodeCiphertext(encodedText string, rows int) string {
	size := len(encodedText)
	if rows == 1 || size <= 1 {
		return encodedText
	}

	cols := size / rows
	posX, posY := 0, 0
	var ret strings.Builder

	for posY < cols {
		index := posX*cols + posY
		ret.WriteByte(encodedText[index])

		posX++
		posY++

		if posX >= rows {
			posX = 0
			posY = posY - rows + 1
		}
	}

	return strings.TrimRight(ret.String(), " ")
}
